modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheAffinityContext.java

```diff
@@ -126,26 +141,50 @@ public synchronized boolean readPartitionsUpdateResponse(PayloadInputChannel ch)
         if (lastTop.get() == null)
             return false;
 
-        ClientCacheAffinityMapping newMapping = ClientCacheAffinityMapping.readResponse(ch);
+        CacheMappingRequest rq0 = rq;
+
+        ClientCacheAffinityMapping newMapping = ClientCacheAffinityMapping.readResponse(ch,
+            new Function<Integer, Function<Integer, ClientPartitionAwarenessMapper>>() {
+                @Override public Function<Integer, ClientPartitionAwarenessMapper> apply(Integer cacheId) {
+                    synchronized (cacheKeyMapperFactoryMap) {
+                        ClientPartitionAwarenessMapperHolder hld = cacheKeyMapperFactoryMap.get(cacheId);
+
+                        // Factory concurrently removed on cache destroy.
+                        if (hld == null || hld.ts == REMOVED_TS)
+                            return null;
+
+                        return hld.factory;
+                    }
+                }
+            }
+        );
+
+        // Clean up outdated factories.
+        rq0.caches.removeAll(newMapping.cacheIds());
+
+        synchronized (cacheKeyMapperFactoryMap) {
+            cacheKeyMapperFactoryMap.entrySet()
+                .removeIf(e -> rq0.caches.contains(e.getKey())
+                    && e.getValue().ts <= rq0.ts);
+        }
+
+        rq = null;
 
         ClientCacheAffinityMapping oldMapping = affinityMapping;
 
-        if (oldMapping == null || newMapping.topologyVersion().compareTo(oldMapping.topologyVersion()) > 0) {
+        if (oldMapping == null || newMapping.compareTo(oldMapping) > 0) {
             affinityMapping = newMapping;
 
+            // Re-request mappings that are out of date.
             if (oldMapping != null)
                 pendingCacheIds.addAll(oldMapping.cacheIds());
 
-            pendingCacheIds.removeAll(newMapping.cacheIds());
```

Comment:
We should exclude cache ids we already aware (with such a change duplicated information will be requested with the next request)

Reply:
Fixed.